src/protocols/ssh/client.c

```diff
@@ -70,8 +70,14 @@ int guac_ssh_client_free_handler(guac_client* client) {
 
     /* Free terminal (which may still be using term_channel) */
     if (ssh_client->term != NULL) {
-        guac_terminal_free(ssh_client->term);
+        /* Close user input pipe to stop reading in ssh_input_thread */
+        close(ssh_client->term->stdin_pipe_fd[1]);
+        close(ssh_client->term->stdin_pipe_fd[0]);
+        ssh_client->term->stdin_pipe_fd[1] = -1;
+        ssh_client->term->stdin_pipe_fd[0] = -1;
+
```

Comment:
Why not just leave the call to guac_terminal_free here?  Doesn't that function do exactly the same thing that you're manually doing here?  It seems like this is a violation of abstraction (I made that term up) here to be manually modifying the data members of the terminal object rather than letting the methods in terminal.c handle that.

Reply:
I agree this looks really ugly. But if I don't close the pipe here, there will be a dead lock in this case.

The thread here will be waiting for ssh_client_thread() by calling pthread_join(ssh_client->client_thread, NULL), ssh_client_thread() will be waiting for ssh_input_thread(). ssh_input_thread() will be blocked by guac_terminal_read_stdin() show below.
https://github.com/sanhex/incubator-guacamole-server/blob/c6cb1910e3580f9c21fc202f824fc55c95a43e49/src/protocols/ssh/ssh.c#L161

The child process of guacd will never exit after disconnecting the ssh connection. I met this problem.. So I manually closed the pipe to stop the ssh_input_thread() mentioned above.

Reply:
So, you're saying that, even with the changes to the `guac_terminal_free()` function, a call to that function, here, would result in the deadlock?

@ceharris @mike-jumper Thoughts on this one?

Reply:
Sorry for the confusion. The changes in this patch to guac_terminal_free() are only for a better practice of preventing unnecessary double closing. It does nothing with the dead lock. The dead lock will happen if I only switch the guac_terminal_free() and pthread_join(). ssh_input_thread() won't be terminated if I don't manually close the pipe here.

Reply:
No need to be sorry - I'm just trying to wrap my head around it, and I'm much more familiar with the guacamole-client code than the server code.  Sometimes I need things explained in small words, repeatedly.

I'm curious what input Mike or Carl have on this - it still makes me a little twitchy to modify these member fds directly in the ssh.c file, but it could be that I'm just being unreasonable.

Reply:
> .. it still makes me a little twitchy to modify these member fds directly in the ssh.c file ...

This makes me twitchy, too. I'd much rather that the handling of file descriptor closure be encapsulated within the code that needs to know about it. I don't like the idea of requiring that all users of the terminal manually close the terminal's internal descriptors.

Reply:
@mike-jumper Will it make more sense if I add a new function e.g. guac_terminal_close_pipe() in terminal/terminal.c, call it here and also in guac_terminal_free() ?

Reply:
No, that would still expose internals, this time within the name of the function itself.

The new ordering with waiting on the thread prior to freeing the terminal is correct, but it seems to me that it's the client thread's dependence on the terminal that's the issue here. The input thread shouldn't be depending on closure of the terminal at all. The various client-specific threads should take into account the client's own stop signal, rather than relying on `guac_terminal` to expand scope and shoulder that burden.

Reply:
The problem here is the pipe is blocking. The call guac_terminal_read_stdin() in ssh_input_thread() will never return when there is no data in the pipe after ssh disconnecting. Another option is setting the pipe fds to non-blocking and tweaking the below callers to read and sleep.

./protocols/telnet/telnet.c:267:    while ((bytes_read = guac_terminal_read_stdin(telnet_client->term, buffer, sizeof(buffer))) > 0) {
./protocols/ssh/ssh.c:161:    while ((bytes_read = guac_terminal_read_stdin(ssh_client->term, buffer, sizeof(buffer))) > 0) {
./terminal/terminal.c:622:    while (guac_terminal_read_stdin(terminal, &in_byte, 1) == 1) {

Reply:
> The problem here is the pipe is blocking. The call guac_terminal_read_stdin() in ssh_input_thread() will never return when there is no data in the pipe after ssh disconnecting.

In that case, I'd suggest something like `guac_terminal_stop()` to provide the needed functionality (forcing reads from `guac_terminal` to fail) without requiring exposure of internals.

Reply:
My first feeling was sending an EOF (maybe defined as -1 on Linux) to the pipe forcing the read to fail. I'm not sure if it will work cos I never try this on a pipe. Any better idea?

Reply:
Well:

1. Keep in mind that guacamole-server is meant to be as POSIX-compatible as possible. Unless something cannot be done without platform-specific code (`move-fd.c` is the only current case of this in the codebase), then it needs to be done without platform-specific code.
2. You can't actually send EOF, as EOF isn't data, but a state. If you want reads to fail from one end of the pipe and return EOF, the way to do that is to close the file descriptor on the other end.

Closing the file descriptors to force reads to fail is fine. The problem is requiring the caller to know that this is what is happening. As long as the function naming and documentation do not require the caller to have such internal knowledge of `guac_terminal`, then things are OK.

It makes sense for the caller to need to know:

* Terminal reads/writes may block (I believe this is already documented).
* Some specific function (`guac_terminal_stop()`? `guac_terminal_close()`?) can be used to forcibly unblock any pending reads/writes and cease all terminal I/O.

It doesn't make sense for the caller to need to know:

* `guac_terminal` contains multiple file descriptors for pipes which are used internally by the read/write functions and should be closed to force those reads/writes to unblock and fail.